modules/core/src/main/java/org/apache/ignite/internal/processors/pool/PoolProcessor.java

```diff
@@ -462,7 +465,39 @@ public PoolProcessor(GridKernalContext ctx) {
             cfg.getQueryThreadPoolSize(),
             cfg.getQueryThreadPoolSize(),
             DFLT_THREAD_KEEP_ALIVE_TIME,
-            new LinkedBlockingQueue<>(),
+            new PriorityBlockingQueue<>(11, new Comparator<Runnable>() {
+                @Override public int compare(Runnable o1, Runnable o2) {
+                    Byte leftPriority = null;
+                    Byte rightPriority = null;
+                    long leftTs = 0;
+                    long rightTs = 0;
+
+                    if (o1 instanceof PriorityWrapper) {
+                        PriorityWrapper left = (PriorityWrapper)o1;
+                        leftPriority = (byte)-left.priority();
+                        leftTs = left.initializedTs();
+                    }
+
+                    if (o2 instanceof PriorityWrapper) {
+                        PriorityWrapper right = (PriorityWrapper)o2;
+                        rightPriority = (byte)-right.priority();
+                        rightTs = right.initializedTs();
+                    }
+
+                    if (leftPriority != null && rightPriority != null) {
+                        int comp = Byte.compare(leftPriority, rightPriority);
+                        return comp == 0 ? Long.compare(leftTs, rightTs) : comp;
+                    }
+                    else if (leftPriority != null || rightPriority != null) {
+                        return leftPriority != null ? leftPriority : rightPriority;
```

Comment:
It doesn't make a sense. You always return a positive number, which means the first argument is less than the second.

Reply:
thanks, this part is totally refactored.